### PR TITLE
feat(compiler): introduce schema for elements

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
@tbosch this is the re-done PR for elements schema support. I believe that I've addressed all the comments from yesterday: we've got a minimal API / set of changes here to support property bindings to work as today. We can build on top of this later on (events, attributes etc.)

Could you PTAL?

File: modules/angular2/src/render/dom/view/proto_view_builder.ts [hunk 1]
```diff
@@ -349,29 +352,25 @@ function buildElementPropertyBindings(protoElement: /*element*/ any, isNgCompone
...
-    var tagName = DOM.tagName(protoElement);
-    var possibleCustomElement = tagName.indexOf('-') !== -1;
-    if (possibleCustomElement && !isNgComponent) {
-      // can't tell now as we don't know which properties a custom element will get
-      // once it is instantiated
-      return true;
+    if (!isNgComponent) {
+      return schemaRegistry.hasProperty(protoElement, binding.property);
     } else {
+      // TODO(pk): here we are checking directive props so shouldn't use DOM facade
```
Comment (tbosch) [inline modules/angular2/src/render/dom/view/proto_view_builder.ts]: Yes and no.
The method `isValidElementPropertyBindings` is used for host properties and regular property bindings. For host properties, we should only check DOM properties. Previously we used the knowledge of whether this as an ng component to assume that it is not a custom element, and therefore could call `DOM.hasProperty()`.

For regular property bindings you are right in that for ng components we should only allow properties of that component.

**Comment by tbosch** [comment]
Looks good, except for the comment above. Could you clean this up?

**Comment by pkozlowski-opensource** [comment]
@tbosch removed checks for directive properties on host elements and landed it. Will follow up with issues for missing features (events, attributes) and PRs targeting custom components.

**Comment by tbosch** [comment]
But you kept the `TODO` in that place / didn't use the schema...
And could you add a comment why you pass in `true` when checking the host properties? It's not clear from the code...

**Comment by pkozlowski-opensource** [comment]
@tbosch yes, will sent PRs with additional tests / comments first thing tomorrow. Just wanted to land this one so it doesn't get bit-rotten.

**Comment by tbosch** [comment]
ok

**Pull request closed without merging**